### ISSUE TITLE
Ctrl-Z can't suspend dlv

### ISSUE THREAD

**Issue by qingyunha** [go-delve/delve#2157]
related to https://github.com/peterh/liner/issues/131

**Comment by aarzilli** [comment]
Why should we do this?

**Comment by qingyunha** [comment]
I usaully use Ctrl-Z and fg for quick-switch and gdb can handle Ctrl-Z.

**Comment by dkegel-sigsci** [comment]
This is a serious usability problem.  I am constantly pausing debuggers etc. to look at source code.

**Comment by qingyunha** [comment]
I made a pull request to liner https://github.com/peterh/liner/pull/149, it may solve this problem.

**Comment by thockin** [comment]
I echo my support - this is painful

**Comment by dkegel-fastly** [comment]
Looks like the way forward is to replace liner with something else, e.g. a fork of liner that has the fix.

**Comment by thockin** [comment]
I am sympathetic to the idea that liner wants to act like a shell.  Without
knowing much about it, it seems like liner needs to pass the SIGTSTP to dlv
rather than eating it.  Liner is not the primary app here - let the app
control signals.

**Comment by derekparker** [comment]
Yes, ultimately this is an issue with liner not passing through SIGTSTP, so ultimately the solution would likely be to maintain our own fork of liner with our own patch applied to "fix" this. Another option would be to move off of liner, however it has some very admirable qualities (e.g. cross-platform, no cgo).

I'm not fully opposed to forking liner within the go-delve org namespace and applying the necessary code changes. I think rebasing to pick up future versions of liner should be relatively painless.

**Comment by derekparker** [comment]
@aarzilli / @polinasok thoughts on forking liner within go-delve org and maintaining that? I think it'd be rather painless, but I'd like to get buy-in from all maintainers before moving forward as opposed to bulldozing this change through.

Also open to other options.

**Comment by aarzilli** [comment]
Upon further consideration my opinion is that we should not implement this, for the following reasons:
1. it will not work on macOS due to limitations with debugserver
2. it will not work on windows due to limitations with how we spawn the target process
3. it would be hard to test because terminal & signals
4. existing "workarounds" (running under gnu screen, tabbed terminals) actually offers a better user experience than relying on the archaic shell job control
I'm not opposed to forking liner (or any other dependency), this doesn't seem like a good reason enough to do it. A year ago I asked "why should we do this" and I don't see a good answer to that here, besides that gdb (first released in 1986) happens to support it.

**Comment by dkegel-fastly** [comment]
Nobody expects it to work on Windows, so that's ok.

What's the debugserver problem?  All we want is to suspend the dlv cli so we can access the terminal, so it's ok if debugserver keeps running.

The workarounds aren't great because users do not expect to have to use them.  It is hard to retrain fingers used to a lifetime of ^Z working on Unix.

**Comment by aarzilli** [comment]
> What's the debugserver problem?

debugserver places the target program on the same process group as itself so signals sent from the controlling terminal of delve will be also sent to the target program.

> The workarounds aren't great because users do not expect to have to use them

I think at this point people who use screen/tmux or a windowed GUI outnumber the people who know about terminal job control by several orders of magnitude.

**Comment by dkegel-fastly** [comment]
Unix/Linux/Mac GUI users who use terminal windows still rely on ^Z, so supporting ^Z is still important.

lldb can handle ^Z; I just tried it with a simple C program, and it seemed to work fine.

It'd be sad if delve was less functional than lldb in this respect.

**Comment by thockin** [comment]
I appreciate how useful and usable delve has become, and how often my gdb
muscle-memory "just works".  Lacking ^Z is exceedingly unfortunate for
those people who expect "archaic shell job control" to be the norm on
archaic (read: Linux) OSes.

> A year ago I asked "why should we do this" and I don't see a good answer
to that here, besides that gdb (first released in 1986) happens to support
it.

Just about EVERY UNIX application supports ^Z because you don't actually
have to do anything to support it.  It seems like something is actively
subverting the "normal" behavior.  NOT supporting it makes delve the
odd-man-out in an otherwise totally natural ecosystem.

**Comment by dkegel-fastly** [comment]
Yeah, the impression the frustrated user gets is that delve is actively interfering with normal keyboard controls.  That is only partly true (it's liner that is doing that), but that's how people react.

**Comment by derekparker** [comment]
> I appreciate how useful and usable delve has become, and how often my gdb muscle-memory "just works". 

Thank you for the kind words about Delve, Tim! It's very, very much appreciated.

> Lacking ^Z is exceedingly unfortunate for those people who expect "archaic shell job control" to be the norm on archaic (read: Linux) OSes.

I completely understand this frustration, as I still very much have the ^Z muscle memory engrained in my workflow.

> A year ago I asked "why should we do this" and I don't see a good answer
> to that here, besides that gdb (first released in 1986) happens to support it. 
> Just about EVERY UNIX application supports ^Z because you don't actually have to do anything to support it. It seems like something is actively subverting the "normal" behavior. NOT supporting it makes delve the odd-man-out in an otherwise totally natural ecosystem.

I agree that the current state of things presents an unexpected and surprising user experience, which is against what Delve aims to provide. The ethos and motto of Delve has always been "get out of the way as much as possible". You're already fighting against the bug you're trying to fix, you shouldn't have to also fight the debugger.

This is a solvable problem. We have the technology ™️. I think making assumptions about a particular developer's workflow is leading us down the wrong path. I agree this is a problem to be solved, so let's move forward under that pretense.

I will start tackling this problem tomorrow, and if that means we fork peterh/liner under the go-delve org to carry the patch necessary to support this than that is what we will do.

**Comment by dkegel-fastly** [comment]
Let me just add that I used delve last night to track something down, and it was a real pleasure to have things just work.
The amount of relearning needed to start using delve is relatively small.

**Comment by thockin** [comment]
Indeed, I find myself reaching for delve far more often than printf() these
days :)

It's also a great way to start to comprehend a new program.

**Comment by polinasok** [comment]
I haven't felt the pain of not being able to Ctrl-Z, but I have found it annoying not to be able to take down a headless server with Ctrl-C in the past. So I can relate to the need to have familiar shortcuts working. It's unfortunate that liner is choosing to not support this because it is not cross-platform. Maybe they will change their mind once they see this in action in dlv.
